drivers/nutdrv_qx.c

```diff
@@ -1911,17 +2036,17 @@ static int	armac_command(const char *cmd, char *buf, size_t buflen)
 	memset(buf, 0, buflen);
 
 	bufpos = 0;
-	while (bufpos + ARMAC_READ_SIZE < buflen) {
+	while (bufpos + read_size + 1 < buflen) {
```

Comment:
This `+1` along 0x3f instead of 0x0f is pretty much the only change I'm not certain won't affect all other versions.

We'd need to just test it on hardware I guess. Or create a database of recorded communications and do a regression testing on this.

Reply:
Well, the `+1` is for ensuring a zero-byte in the end (of buffer to be provided big enough by the caller).

Reply:
Thinking a bit more about this, it may be prudent to check that `buflen > read_size` in advance: if the caller's buffer is too small to fit even the first read eventually (if `usb_interrupt_read()` would return the full requested `read_size` and not fewer bytes for some reason), this is more of a programmer error to `upsdebugx()` as such than a device/protocol error. I think it can help long-term troubleshooting as this code evolves.

Then there's a chance that we expect to read say 64 bytes (due to `if use_interrupt ...` above) but only get say 6 in reality (guessing here, no idea if this reality is possible - WDYT?), and the caller's buffer is big enough for 6 but not for 64. With current while-clause, we would not even bother trying a query. Is this a correct thing to do?

Just trying to think of things that could go wrong, so might brainstorm up some garbage ideas here :)

Reply:
> it may be prudent to check that buflen > read_size in advance

If I understand it right we already have this check - we won't enter loop at all in that case `while (bufpos + read_size + 1 < buflen)`. And bellow the loop we emit the error for this specific case `upsdebugx(2, "Protocol error, too much data read.");`. It still had hardcoded read_size that was fixed via 66ff781

> Then there's a chance that we expect to read say 64 bytes (due to if use_interrupt ... above) but only get say 6 in reality 

At least in my case I do always get 64 bytes from `usb_interrupt_read`. From them we have only 47 meaningful bytes that will be copied to `buf`. This is handled via `bytes_available = (unsigned char)tmpbuf[0] & 0x3f;`. In theory this function could be changed to work when caller's buffer would be less than 64 bytes. But then we are risking that all data we read from interrupt won't fit to `buf`. I'm not sure which way is better?